Definer.fs

```diff
@@ -0,0 +1,281 @@
+       : Result<Func<Expr<'Var>>, Error> =
+        List.map2
+            (curry
+                 (function
+                  | UnifyInt _ | UnifyBool _ -> ok ()
+                  | UnifyFail (fp, dp) -> fail (TypeMismatch (dp, typeOf fp))))
+            func.Params
+            defn.Params
+        |> collect
+        |> lift (fun _ -> func)
+
+    /// <summary>
+    ///     Look up <c>func</c> in <c>_arg1</c>.
+    ///
+    ///     <para>
+    ///         This checks that the use of <c>func</c> agrees on the number of
+    ///         parameters, but not necessarily types.  You will need to add
+    ///         type checking if needed.
+    ///     </para>
+    /// </summary>
+    /// <param name="func">The func to look up in <c>definer</c>.</param>
+    /// <param name="definer">
+    ///     A definer mapping <c>Func</c>s to some definition.
+    /// </param>
+    /// <typeparam name="Defn">The type of definer definitions.</typeparam>
+    /// <returns>
+    ///     A Chessie result, where the <c>ok</c> value is an <c>Option</c>
+    ///     containing the pair of
+    ///     prototype func and definition, and the failure value is a
+    ///     <c>Starling.Core.Definer.Error</c>.  If the <c>ok</c> value is
+    ///     <c>None</c>, it means no (valid or otherwise) definition exists.
+    /// </returns>
+    let lookup (func : Func<_>) (definer : FuncDefiner<'Defn>)
+      : Result<(DFunc * 'Defn) option, Error> =
+        // First, try to find a func whose name agrees with ours.
+        let defSeq = toSeq definer
+        let defn = Seq.tryFind (fun (dfunc, _) -> dfunc.Name = func.Name) defSeq
+        checkParamCount func defn
+
+    /// <summary>
+    ///     Look up <c>func</c> in <c>_arg1</c>, and, if it exists, typecheck
+    ///     it against its definition.
+    /// </summary>
+    /// <param name="func">
+    ///     The func to look up in <c>_arg1</c>.
+    /// </param>
+    /// <param name="definer">
+    ///     An associative sequence mapping <c>Func</c>s to some definition.
+    /// </param>
+    /// <typeparam name="Defn">The type of definer definitions.</typeparam>
+    /// <returns>
+    ///     A Chessie result, where the <c>ok</c> value is an <c>Option</c>
+    ///     containing the pair of
+    ///     prototype func and definition, and the failure value is a
+    ///     <c>Starling.Core.Definer.Error</c>.  If the <c>ok</c> value is
+    ///     <c>None</c>, it means no (valid or otherwise) definition exists.
+    /// </returns>
+    let lookupWithTypeCheck
+      (func : Func<Expr<_>>) (definer : FuncDefiner<'Defn>)
+      : Result<(DFunc * 'Defn) option, Error> =
+        let unchecked = lookup func definer
+
+        bind
+            (function
+             | None -> ok None
+             | Some (dfunc, defn) ->
+                 lift
+                     (fun _ -> Some (dfunc, defn))
+                     (checkParamTypes func dfunc))
+            unchecked
+
+
+module ViewDefiner =
```

Comment:
After messing around for several hours trying to add one and use it, I've come to the conclusion that it's a waste of time :P  We never use the definitions here directly, we always convert to funcdefiner first, and all we need the viewdefiner for is checking whether definitions _exist_, not what they are.